### PR TITLE
ES6 Map, Set and initial Iterators

### PR DESCRIPTION
**Pull request by lahma**
First implementation where most tests pass and one can use new Map(), new Set() and use them pretty much as expected. Some shortcomings listed in skipped.json.

### CODE REVIEW

File: Jint/Native/Function/FunctionInstance.cs [hunk 1]
```diff
@@ -15,23 +15,35 @@ public abstract class FunctionInstance : ObjectInstance, ICallable
         private const string PropertyNameLength = "length";
         private const int PropertyNameLengthLength = 6;
         protected PropertyDescriptor _length;
+
+        private const string PropertyNameName = "name";
+        private const int PropertyNameNameLength = 4;
```
Comment (sebastienros) [inline Jint/Native/Function/FunctionInstance.cs]: I assume for performance reasons. Isn't the `==` operator on string cleaver enough to already apply this?

File: Jint/Native/JsValue.cs [hunk 2]
```diff
@@ -142,6 +144,37 @@ public ArrayInstance AsArray()
...
+                return null;
+            }
+
+            if (!(value is ICallable callable))
+            {
+                ExceptionHelper.ThrowArgumentException("The value is not iterable");
+                return null;
+            }
+
+            var obj = (ObjectInstance) callable.Call(this, System.Array.Empty<JsValue>());
```
Comment (sebastienros) [inline Jint/Native/JsValue.cs]: There is a `Arguments.Empty` field for this (same thing in the end, just more meaningful)

File: Jint/Native/Array/IArrayLike.cs [hunk 3]
```diff
@@ -0,0 +1,7 @@
+﻿namespace Jint.Native.Array
+{
+    internal interface IArrayLike
```
Comment (sebastienros) [inline Jint/Native/Array/IArrayLike.cs]: Looking at the spec, an "array-like" object is from TypedArray, and here only String implements it. 
https://www.ecma-international.org/ecma-262/6.0/#sec-typedarray-objects

Also in the Array.from you don't use Iterator at all, is that intentionally left for later?
https://www.ecma-international.org/ecma-262/6.0/#sec-array.from

Comment (sebastienros) [reply Jint/Native/Array/IArrayLike.cs]: I see that what they call "array-like" is any object that has a "length" property and can have properties like `'0'`, ... So yes String is an array-like object. But maybe the interface should actually expose the `Length` and `this[int]` members? Might be easier to implement the spec that expects array-like objects.

File: Jint/Native/Function/FunctionInstance.cs (hunk 1 above)
Comment (lahma) [reply Jint/Native/Function/FunctionInstance.cs]: I measured it with BDN at it showed to be a bit faster when we don't need a null check before we do length check, so a real micro-optimization that might be something to remove at some point.

File: Jint/Native/Array/IArrayLike.cs (hunk 3 above)
Comment (lahma) [reply Jint/Native/Array/IArrayLike.cs]: I started to check Array items and was overwhelmed with the stuff that is needed. So I was thinking maybe process this part first and when PR lands continue with adding missing bits to strings and Array.

**Pull request merged**